### PR TITLE
Incremental BMC for CEGIS

### PR DESCRIPTION
**Pull request by peterschrammel**
thanks to @polgreen for fixing the merge problems

### CODE REVIEW

File: src/big-int/bigint-test.cc [hunk 1]
```diff
@@ -116,9 +116,9 @@ In its original version it was built by
 
 at
 
-    Institut f�r Software- und Systemtechnik der Fraunhofer-Gesellschaft
+    Institut für Software- und Systemtechnik der Fraunhofer-Gesellschaft
     (Fraunhofer Institute for Software Engineering and Systems Engineering)
-    Kurstra�e 33
+    Kurstraße 33
   D-10117 Berlin
```
Comment (kroening) [inline src/big-int/bigint-test.cc]: What is this change?

File: src/cbmc/cbmc_parse_options.cpp [hunk 2]
```diff
@@ -430,6 +460,50 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
+  if(cmdline.isset(opt1) && !cmdline.isset(opt2))
+  {
+    error() << "--" << opt1 << " can only be used with --" << opt2 << eom;
+    return true;
+  }
+  return false;
+}
+
+/*******************************************************************\
+
```
Comment (kroening) [inline src/cbmc/cbmc_parse_options.cpp]: Please make separate pull request for the two above; this seems unrelated to incremental.

File: src/cbmc/symex_bmc_incremental.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    if(loop_cond.checked_function)
+    {
+      loop_cond.checked_function=false;
+      return false;
+    }
+
+#if 1
+    if(options.get_bool_option("magic-numbers") &&
+         magic_numbers.find(unwind)==magic_numbers.end())
+    {
```
Comment (kroening) [inline src/cbmc/symex_bmc_incremental.cpp]: This seems to be obscure.

File: src/goto-symex/symex_target_equation.cpp [hunk 4]
```diff
@@ -32,7 +33,7 @@ Function: symex_target_equationt::symex_target_equationt
 \*******************************************************************/
 
 symex_target_equationt::symex_target_equationt(
-  const namespacet &_ns):ns(_ns)
+  const namespacet &_ns):is_incremental(false),ns(_ns),io_count(0)
 {
```
Comment (kroening) [inline src/goto-symex/symex_target_equation.cpp]: spaces missing

File: src/goto-symex/symex_target_equation.cpp [hunk 5]
```diff
@@ -645,12 +646,15 @@ Function: symex_target_equationt::convert_assignments
 \*******************************************************************/
 
 void symex_target_equationt::convert_assignments(
-  decision_proceduret &decision_procedure) const
+  decision_proceduret &decision_procedure)
 {
```
Comment (kroening) [inline src/goto-symex/symex_target_equation.cpp]: Why did the const get removed?

File: src/util/string_utils.cpp [hunk 6]
```diff
@@ -8,6 +8,7 @@ Author: Daniel Poetzl
```
Comment (kroening) [inline src/util/string_utils.cpp]: The changes to this file seem unrelated.

File: src/cbmc/cbmc_parse_options.cpp (hunk 2 above)
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: It's only used for the incremental options here, but it might be useful for other option consistency checks too -- I put it into a separate commit.

File: src/goto-symex/symex_target_equation.cpp (hunk 5 above)
Comment (peterschrammel) [inline src/goto-symex/symex_target_equation.cpp]: I have to update step.converted to remember which steps are already in the solver.

File: src/cbmc/cbmc_parse_options.cpp [hunk 7]
```diff
@@ -1138,6 +1180,13 @@ void cbmc_parse_optionst::help()
     " --unwind nr                  unwind nr times\n"
     " --unwindset L:B,...          unwind loop L with a bound of B\n"
     "                              (use --show-loops to get the loop IDs)\n"
+    " --incremental                check after each unwinding\n"
+    " --incremental-check L        check after each unwinding of loop L\n"
+    " --unwind-min nr              start incremental check after nr unwindings\n"
+    " --unwind-max nr              stop incremental check after nr unwindings\n"
+    " --earliest-loop-exit         stop unwinding as soon as possible\n"
+    " --ignore-assertions-before-unwind-min\n"
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: No explanatory text?

File: src/cbmc/cbmc_parse_options.cpp [hunk 8]
```diff
@@ -213,6 +226,24 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
+  {
+    options.set_option("refine", true);
+    options.set_option("refine-arrays", true);
+    options.set_option(
+      "incremental-check", cmdline.get_value("incremental-check"));
+  }
+  if(cmdline.isset("earliest-loop-exit"))
+    options.set_option("earliest-loop-exit", true);
+  if(cmdline.isset("magic-numbers"))
+    options.set_option("magic-numbers", true);
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: This option is undocumented.

File: src/cbmc/fault_localization.cpp [hunk 9]
```diff
@@ -388,6 +388,7 @@ safety_checkert::resultt fault_localizationt::stop_on_fail()
 
     return safety_checkert::ERROR;
   }
+  assert(false);
```
Comment (tautschnig) [inline src/cbmc/fault_localization.cpp]: What is this necessary (the `switch` has a `default`)?

File: src/cbmc/symex_bmc_incremental.cpp [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+  magic_numbers.insert(6);
+  magic_numbers.insert(12);
+  magic_numbers.insert(17);
+  magic_numbers.insert(21);
+  magic_numbers.insert(40);
+  magic_numbers.insert(60);
+  magic_numbers.insert(80);
+  magic_numbers.insert(100);
+  magic_numbers.insert(200);
+  magic_numbers.insert(1025);
```
Comment (tautschnig) [inline src/cbmc/symex_bmc_incremental.cpp]: I'm scared - magic numbers without a single comment.

File: src/cbmc/symex_bmc_incremental_one_loop.cpp [hunk 11]
```diff
@@ -0,0 +1,132 @@
+/*******************************************************************\
+
+Module: Incremental Bounded Model Checking for ANSI-C
+
+Author: Peter Schrammel, Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+#include <limits>
+#include <iostream>
```
Comment (tautschnig) [inline src/cbmc/symex_bmc_incremental_one_loop.cpp]: Not used (debugging/#if0 only)

File: src/cbmc/symex_bmc_incremental_one_loop.cpp [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  loop_limitst &this_thread_limits=
+    thread_loop_limits[state.source.thread_nr];
+  if(incr_loop_id=="" &&
+     this_thread_limits.find(id)==this_thread_limits.end() &&
+     loop_limits.find(id)==loop_limits.end())
+  {
+    std::cout << "not statically unwound" << std::endl;
+    // not a statically unwound loop when incremental
+  }
+#endif
```
Comment (tautschnig) [inline src/cbmc/symex_bmc_incremental_one_loop.cpp]: Admittedly this function has a lot of #if0

File: src/goto-symex/goto_symex_state.h [hunk 13]
```diff
@@ -21,6 +21,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "symex_target.h"
 
+#define SYMEX_CONTINUATION_CHECK "symex_continuation_check"
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.h]: I would suggest to add an irep_id ID_C_symex_... (not sure about the "continuation" part, see above) as this will inevitably end up in the string table.

File: src/goto-symex/symex_function_call.cpp [hunk 14]
```diff
@@ -71,7 +71,7 @@ void goto_symext::parameter_assignments(
 
   // iterates over the types of the parameters
   for(code_typet::parameterst::const_iterator
-      it2=parameter_types.begin();
+        it2=parameter_types.begin();
```
Comment (tautschnig) [inline src/goto-symex/symex_function_call.cpp]: That looks wrong

File: src/goto-symex/symex_function_call.cpp [hunk 15]
```diff
@@ -544,9 +549,9 @@ void goto_symext::return_assignment(statet &state)
       if(!base_type_eq(assignment.lhs().type(),
                        assignment.rhs().type(), ns))
         throw "goto_symext::return_assignment type mismatch at "+
-              instruction.source_location.as_string()+":\n"+
-              "assignment.lhs().type():\n"+assignment.lhs().type().pretty()+"\n"+
-              "assignment.rhs().type():\n"+assignment.rhs().type().pretty();
+          instruction.source_location.as_string()+":\n"+
+          "assignment.lhs().type():\n"+assignment.lhs().type().pretty()+"\n"+
+          "assignment.rhs().type():\n"+assignment.rhs().type().pretty();
```
Comment (tautschnig) [inline src/goto-symex/symex_function_call.cpp]: There seem to be some whitespace changes that could be considered debatable.

File: src/goto-symex/symex_goto.cpp [hunk 16]
```diff
@@ -171,8 +185,35 @@ void goto_symext::symex_goto(statet &state)
...
+  bool is_function,
+  statet &state,
+  const exprt &cond,
+  unsigned unwind)
+{
+  // dummy implementation
+  return false;
 }
 
+
```
Comment (tautschnig) [inline src/goto-symex/symex_goto.cpp]: Why another blank line?

File: src/goto-symex/symex_main.cpp [hunk 17]
```diff
@@ -143,43 +143,48 @@ void goto_symext::rewrite_quantifiers(exprt &expr, statet &state)
...
-    symex_step(goto_functions, state);
+    if(symex_step(goto_functions, state))
+      return false;
 
     // is there another thread to execute?
     if(state.call_stack().empty() &&
        state.source.thread_nr+1<state.threads.size())
     {
       unsigned t=state.source.thread_nr+1;
-      //std::cout << "********* Now executing thread " << t << std::endl;
```
Comment (tautschnig) [inline src/goto-symex/symex_main.cpp]: Why does this line need to go away?

File: src/goto-symex/symex_target_equation.h [hunk 18]
```diff
@@ -244,23 +248,39 @@ class symex_target_equationt:public symex_targett
   unsigned count_assertions() const
   {
     unsigned i=0;
-    for(SSA_stepst::const_iterator
-        it=SSA_steps.begin();
-        it!=SSA_steps.end(); it++)
-      if(it->is_assert()) i++;
+    for(const auto & it : SSA_steps)
```
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.h]: Nit: no space after `&`

File: src/cbmc/bmc.h [hunk 19]
```diff
@@ -149,7 +146,12 @@ class bmct:public safety_checkert
   friend class fault_localizationt;
 
 private:
-  symex_bmct &symex;
+  // We cannot use a reference member here
+  // because the initialization sequence would not be clean (Clang complains)
```
Comment (tautschnig) [inline src/cbmc/bmc.h]: What exactly does Clang complain about? I fail to see the problem with the original code...

File: src/cbmc/bmc.cpp [hunk 20]
```diff
@@ -69,6 +69,9 @@ Function: bmct::error_trace
 
 void bmct::error_trace()
 {
+  if(options.get_bool_option("stop-when-unsat"))
+    return;
+
```
Comment (polgreen) [inline src/cbmc/bmc.cpp]: The following regression tests use "stop-when-unsat":
cbmc-incr/induction1
cbmc-incr-oneloop/induction1
cbmc-incr-oneloop/induction2

File: src/cbmc/symex_bmc_incremental.cpp (hunk 10 above)
Comment (polgreen) [inline src/cbmc/symex_bmc_incremental.cpp]: There's a comment now: https://github.com/peterschrammel/cbmc/blob/incremental-bmc-rebased/src/cbmc/symex_bmc_incremental.cpp#L47

File: src/cbmc/symex_bmc_incremental_one_loop.cpp (hunk 12 above)
Comment (polgreen) [inline src/cbmc/symex_bmc_incremental_one_loop.cpp]: does that mean the above comment is ok..?

File: src/goto-symex/symex_goto.cpp (hunk 16 above)
Comment (polgreen) [inline src/goto-symex/symex_goto.cpp]: it's not there now

File: src/cbmc/all_properties.cpp [hunk 21]
```diff
@@ -97,7 +97,8 @@ safety_checkert::resultt bmc_all_propertiest::operator()()
       it!=bmc.equation.SSA_steps.end();
       it++)
   {
-    if(it->is_assert())
+    if(it->is_assert() &&
+       it->comment!=SYMEX_CONTINUATION_CHECK)
```
Comment (polgreen) [inline src/cbmc/all_properties.cpp]: I don't know the answer to this.. @peterschrammel, can you help me?

**Comment by polgreen** [comment]
The linter failures on this pull request are:

- src/cbmc/cbmc_parse_options.cpp:1138. The line is as-is in the master, and was last changed in this commit, so I'm not sure how to fix this: https://github.com/peterschrammel/cbmc/commit/f98c61a844fc416ad6f0ef06b727490a6de5d6d5 
- src/goto-symex/symex_main.cpp:283: This line introduces an "assert". The rest of the file contains other asserts. Should I fix this in a separate pull request?

The AppVeyor build fails, for regression tests in cbmc-incr-oneloop that all pass on travis. Is it possible to download the main.out files for the regression tests?

**Comment by tautschnig** [comment]
> src/cbmc/cbmc_parse_options.cpp:1138. The line is as-is in the master, and was last changed in this commit, so I'm not sure how to fix this: peterschrammel/cbmc@f98c61a

I'm not sure why the linter is complaining about line 1138 - the tab is in line 1116, which is introduced in this commit series.

**Comment by tautschnig** [comment]
> The AppVeyor build fails, for regression tests in cbmc-incr-oneloop that all pass on travis. Is it possible to download the main.out files for the regression tests?

You may want to change the Makefiles in those regression test directories to match what other directories do. In particular, the use of `failed-tests-printer.pl` would make sure more information is provided in the output.

**Comment by polgreen** [comment]
The makefile in regression/cbmc-incr-oneloop now matches the makefiles in the other regression test folders. 

The test failures on AppVeyor are odd and I think spurious; cbmc is giving what I think are the correct results, it matches the output on MacOS which passes the regression tests. AppVeyor is reporting failed test.desc lines that appear to be present in the cbmc output, e.g., in the following I don't understand why VERIFICATION FAILED is a failed test.desc line, and why the EXIT and SIGNAL aren't in the output from failed-tests-printer.pl

> ** Results:
> [main.assertion.1] assertion x < 15: SUCCESS
> [main.unwind.0] unwinding assertion loop 0: FAILURE
> ** 1 of 2 failed (2 iterations)
> VERIFICATION FAILED
> Failed test.desc lines:
> ^EXIT=10$ [FAILED]
> ^SIGNAL=0$ [FAILED]
> ^VERIFICATION FAILED$ [FAILED]
> make[1]: *** [test] Error 1

**Comment by tautschnig** [comment]
@polgreen Could you please rebase such as to re-trigger AppVeyor runs? Arguably there is something fishy here, and I'd like to get to the bottom of it...

**Comment by polgreen** [comment]
Yes, will do the rebase tomorrow. 

I spoke to Daniel about the AppVeyor problem and he said he thinks it is probably something to do with perl on AppVeyor. 

I've emailed Vlastimil last week to get an RDP password for AppVeyor, which would help to work out whether it is a problem with perl, but got no response yet.

**Comment by tautschnig** [comment]
Getting back to this one after quite a while: it should be most insightful to compare the tool outputs on cbmc-incr-oneloop/induction1 and cbmc-incr-oneloop/induction2 - these appear to have the byte-by-byte same test.desc files, but only one of them fails. Such a comparison should even be doable on Linux/Mac OS.

**Comment by peterschrammel** [comment]
Replaced by #4361
